radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/PendingOutboundChannelsManager.java

```diff
@@ -128,37 +128,43 @@ public PendingOutboundChannelsManager(
   }
 
   private void handlePeerConnected(PeerConnected peerConnected) {
+    final var channel = peerConnected.channel();
+    final Optional<CompletableFuture<PeerChannel>> channelFutureOpt;
     synchronized (lock) {
-      final var channel = peerConnected.channel();
-      final var maybeFuture = this.pendingChannels.remove(channel.getRemoteNodeId());
-      if (maybeFuture != null) {
-        maybeFuture.complete(channel);
-      }
+      channelFutureOpt =
```

Comment:
Why this fixes the deadlock? Is there another code running in this same thread preventing the Future of being completed?

Reply:
Future's `onComplete` callbacks happen on the thread that calls the `.complete` method (which still holds the lock). The fix is to release the lock before calling the `.complete` method. See a more detailed write up in the Jira ticket.